explainer.md

```diff
@@ -48,218 +49,298 @@ Provide web apps with efficient access to built-in (software and hardware) media
+-   A **VideoDecoder** decodes EncodedVideoChunks to produce VideoFrames.
 
-A **VideoTrackReader** converts a MediaStreamTrack into a ReadableStream of VideoFrame.
 
-A **VideoEncoder** is a TransformStream from VideoFrame to EncodedVideoChunk.
+WebCodecs will also define mechanisms for importing content from getUserMedia().
 
-A **VideoDecoder** is a TransformStream from EncodedVideoChunk to VideoFrame.
+-   An **AudioTrackReader** converts an audio MediaStreamTrack into a ReadableStream of AudioPacket.
 
-A **VideoTrackWriter** converts a WritableStream of VideoFrame into a MediaStreamTrack.
+-   A **VideoTrackReader** converts a video MediaStreamTrack into a ReadableStream of VideoFrame.
 
 ## Examples
-### Example of video rendering to Canvas for low-latency live streaming or cloud gaming
+### Example of video rendering to Canvas for extremely low-latency streaming (e.g. cloud gaming)
 
 ```javascript
-// App defines a simple WritableSink that renders new frames as soon as they arrive.
-class CanvasRendererSink {
-  constructor(canvas) {
-    this._context = canvas.getContext('bitmaprenderer');
-  }
+// The app provides error handling.
+function onDecoderError(error) { ... }
 
-  write(videoFrame) {
-    _context.transferFromImageBitmap(videoFrame.image);
-    return Promise.resolve();
-  }
-}
+// App provides stream of encoded chunks to decoder.
+function streamEncodedChunks(decodeCallback) { ... }
 
-// The app provides ReadableStreams of encoded video
-// in the form of byte arrays defined by a codec such as vp8
-// (not in a media container such as mp4 or webm).
-const {encodedVideo} = ...;
-
-// App sets up canvas on the page for use by the rendering sink.
-const canvas = document.getElementById("canvas");
+// The document contains a canvas for displaying VideoFrames.
+const canvasElement = document.getElementById("canvas");
+const canvasContext = canvasElement.getContext('bitmaprenderer');
 
-// Wrap the rendering sink in a writable stream so it can be connected to the graph.
-const renderingStream = new WritableStream(new CanvasRendererSink(canvas));
+function paintFrameToCanvas(videoFrame) {
+  // Paint every video frame ASAP for lowest latency.
+  canvasContext.transferFromImageBitmap(videoFrame.image);
+}
 
-// Construct the decoder and connect the pipes!
-const videoDecoder = new VideoDecoder({codec: 'vp8'});
-encodedVideo.pipeThrough(videoDecoder).pipeTo(renderingStream);
+const videoDecoder = new VideoDecoder({
+  output: paintFrameToCanvas,
+  error: onDecoderError
+});
 
-```
+videoDecoder.configure({codec: 'vp8'}).then((supported) => {
+  // More elaborate fallback recommended.
+  if (!supported)
+    return;
 
-### Example of decode and rendering via TrackWriters (also for low-latency live streaming or cloud gaming)
+  // The app fetches VP8 chunks, feeding one at time to the
+  // provided callback for decoding.
+  streamEncodedChunks(function(newChunk) {
+    videoDecoder.decode(newChunk);
```

Comment:
I don't know if this will work, we don't guarantee that each decode will complete without calling flush() or reset().

The simplest backpressure is none; just feed all the input. I'd recommend that with a comment that a real app must handle backpressure too.

The expected implementation is a feed() function that pushes chunks as long as there is input available and the backpressure signal is low. (Maybe we need a backpressure changed event/callback?)

Reply:
> I don't know if this will work, we don't guarantee that each decode will complete without calling flush() or reset().

AFAIK I'm not relying on that guarantee? 

> The simplest backpressure is none; just feed all the input. I'd recommend that with a comment that a real app must handle backpressure too.

This is the intended meaning of the code I've written. I'm not sure what meaning you've taken, but LMK if I can name/comment to improve clarity. I've added a comment that real app must also handle backpressure (I guess in the low latency case that would mean Reset() and adapting down to a lower resolution).

Reply:
Sorry, I must have misread.